climada_petals/entity/exposures/osm_dataloader.py

```diff
@@ -0,0 +1,346 @@
+        ways. Assemble ways into gdfs. Keep track of which nodes
+        and ways have been "used up" already
+
+        Parameters
+        ---------
+        result : overpy.Overpass result object
+        nodes_avail : list
+            node-ids that have not yet been used for relation and way
+            construction and are hence available for node constructions
+
+        Returns
+        -------
+        gdf_nodes : gpd.GeoDataFrame
+            gdf with nodes that were assembled from result object
+        """
+        data_geom = []
+        data_id = []
+        data_tags = []
+
+        for node in result.nodes:
+            if node.id in nodes_avail:
+                data_geom.append(shapely.geometry.Point(node.lat, node.lon))
+                data_id.append(node.id)
+                data_tags.append(node.tags)
+
+        gdf_nodes = gpd.GeoDataFrame(
+            data={'osm_id': data_id,'geometry': data_geom, 'tags':data_tags})
+
+        return gdf_nodes
+
+    def _update_availability(self, full_set, to_remove):
+        """
+        update id availabilities from whole result set after using up some for
+        geometry construction
+        """
+        return [item for item in full_set if item not in to_remove]
+
+    def _assemble_results(self, result, closed_lines_are_polys=True):
+        """
+        assemble an overpass result object with results, ways, nodes etc. from the
+        format-specific structure into one geodataframe
+
+        Parameters
+        ---------
+        result : overpy.Overpass result object
+        closed_lines_are_polys : bool
+                whether closed lines are polygons. Default is True
+
+        Returns
+        ------
+        gdf_results : gpd.GeoDataFrame
+            Result-gdf from the overpass query.
+        """
+        gdf_results = gdf_ways = gdf_nodes = gpd.GeoDataFrame(
+            columns=['osm_id','geometry','tags'])
+        nodes_avail = result.node_ids
+        ways_avail = result.way_ids
+
+        if len(result.relations) > 0:
+            nodes_taken, ways_taken, gdf_rels = self._assemble_from_relations(result)
+            gdf_results = gdf_results.append(gdf_rels)
+            nodes_avail = self._update_availability(nodes_avail, nodes_taken)
+            ways_avail = self._update_availability(ways_avail, ways_taken)
+
+        if len(ways_avail) > 0:
+            nodes_taken, gdf_ways = self._assemble_from_ways(
+                result,  ways_avail, closed_lines_are_polys)
+            gdf_results = gdf_results.append(gdf_ways)
+            nodes_avail = self._update_availability(nodes_avail, nodes_taken)
+
+        if len(nodes_avail) > 0:
+            gdf_nodes = self._assemble_from_nodes(result, nodes_avail)
+            gdf_results =  gdf_results.append(gdf_nodes)
+
+        if len(result.nodes) == 0:
+            LOGGER.warning('empty result gdf returned.')
+
+        return gdf_results.reset_index(drop=True)
+
+    def _osm_geoms_to_gis(self, gdf):
```

Comment:
This method should only take the geometry as parameter, because it does not access the rest of the dataframe. Then it can be called like
```python
geo_new = _osm_geoms_to_gis(gdf.geometry)
```